server/src/main/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommand.java

```diff
@@ -0,0 +1,556 @@
+                    }
+                    if (translogStatus != CleanStatus.CLEAN) {
+                        loseDataDetailsBanner(terminal, translogCleanStatus);
+                    }
+                    terminal.println("            WARNING:              YOU MAY LOSE DATA.");
+                    terminal.println("-----------------------------------------------------------------------");
+
+
+                    confirm("Continue and remove corrupted data from the shard ?", terminal);
+
+                    if (indexStatus != CleanStatus.CLEAN) {
+                        removeCorruptedLuceneSegmentsAction.execute(terminal, shardPath, indexDir,
+                            writeIndexLock, printStream, verbose);
+                    }
+
+                    if (translogStatus != CleanStatus.CLEAN) {
+                        truncateTranslogAction.execute(terminal, shardPath, indexDir);
+                    }
+                } catch (LockObtainFailedException lofe) {
+                    final String msg = "Failed to lock shard's directory at [" + indexPath + "], is Elasticsearch still running?";
+                    terminal.println(msg);
+                    throw new ElasticsearchException(msg);
+                }
+
+                final CleanStatus indexStatus = indexCleanStatus.v1();
+                final CleanStatus translogStatus = translogCleanStatus.v1();
+
+                // newHistoryCommit obtains its own lock
+                addNewHistoryCommit(indexDir, terminal, translogStatus != CleanStatus.CLEAN);
+                newAllocationId(environment, shardPath, terminal);
+                if (indexStatus != CleanStatus.CLEAN) {
+                    dropCorruptMarkerFiles(terminal, indexDir, indexStatus == CleanStatus.CLEAN_WITH_CORRUPTED_MARKER);
+                }
+            }
+        }
+    }
+
+    private Directory getDirectory(Path indexPath) {
+        Directory directory;
+        try {
+            directory = FSDirectory.open(indexPath, NativeFSLockFactory.INSTANCE);
+        } catch (Throwable t) {
+            throw new ElasticsearchException("ERROR: could not open directory \"" + indexPath + "\"; exiting");
+        }
+        return directory;
+    }
+
+    protected void addNewHistoryCommit(Directory indexDirectory, Terminal terminal, boolean updateLocalCheckpoint) throws IOException {
+        final String historyUUID = UUIDs.randomBase64UUID();
+
+        terminal.println("Marking index with the new history uuid : " + historyUUID);
+        // commit the new history id
+        final IndexWriterConfig iwc = new IndexWriterConfig(null)
+            // we don't want merges to happen here - we call maybe merge on the engine
+            // later once we stared it up otherwise we would need to wait for it here
+            // we also don't specify a codec here and merges should use the engines for this index
+            .setCommitOnClose(false)
```

Comment:
We need to set the soft-deletes field when opening an IndexWriter. Please see: https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/translog/TruncateTranslogCommand.java#L181

Reply:
thanks